fix: PopulateScenePlacedObjects only checks IsSceneObject for null [MTT-3041]

NoelStephensUnity opened the pull request:
This fixes the issue when using a single scene containing both a NetworkManager and one or more in-scene placed NetworkObjects where a client could disconnect but not unload the scene and reconnect only to receive a soft synchronization error.
The fix is to check for both HasValue and Value being true in order to account for already previously instantiated NetworkObjects.

[MTT-3041](https://jira.unity3d.com/browse/MTT-3041)

## Changelog
- Fixed: in-scene placed NetworkObjects not being found/ignored after a client disconnects and then reconnects.

## Testing and Documentation
- includes a test to verify the changes

NoelStephensUnity committed:
fix

MTT-3041
This fixes the issue when using a single scene containing both a NetworkManager and one or more in-scene placed NetworkObjects where a client could disconnect but not unload the scene and reconnect only to receive a soft synchronization error.
The fix is to check for both HasValue and Value being true in order to account for already previously instantiated NetworkObjects.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Update CHANGELOG.md

MTT-3041 changelog update

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-3041-populatesceneplacedobjects-sceneobject-check-logic

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1858,8 +1858,8 @@ private void PopulateScenePlacedObjects(Scene sceneToFilterBy, bool clearScenePl
-                // We check to make sure the NetworkManager instance is the same one to be "NetcodeIntegrationTestHelpers" compatible and filter the list on a per scene basis (additive scenes)
-                if (networkObjectInstance.IsSceneObject == null && networkObjectInstance.NetworkManager == m_NetworkManager && networkObjectInstance.gameObject.scene == sceneToFilterBy &&
+                // We check to make sure the NetworkManager instance is the same one to be "NetcodeIntegrationTestHelpers" compatible and filter the list on a per scene basis (for additive scenes)
+                if ((!networkObjectInstance.IsSceneObject.HasValue || networkObjectInstance.IsSceneObject.Value) && networkObjectInstance.NetworkManager == m_NetworkManager && networkObjectInstance.gameObject.scene == sceneToFilterBy &&
```
```suggestion
                if (networkObjectInstance.IsSceneObject != false && networkObjectInstance.NetworkManager == m_NetworkManager && networkObjectInstance.gameObject.scene == sceneToFilterBy &&
```

NoelStephensUnity committed:
update

cleaned up PopulateScenePlacedObjects for clarity.
reduce the networkObjectInstance.IsSceneObject check.
made PopulateScenePlacedObjects internal for testing purposes.

NoelStephensUnity committed:
test

MTT-3041
This test verifies the changes made to PopulateScenePlacedObjects

NoelStephensUnity committed:
style

Adding some comments

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1843,7 +1843,7 @@ internal void MoveObjectsToDontDestroyOnLoad()
         /// distinguish between duplicate in-scene placed NetworkObjects
         /// </summary>
-        private void PopulateScenePlacedObjects(Scene sceneToFilterBy, bool clearScenePlacedObjects = true)
+        internal void PopulateScenePlacedObjects(Scene sceneToFilterBy, bool clearScenePlacedObjects = true)
```
Do we really need to use the InternalsVisibleTo hack to test this? What we care about is that in a Stop->Start scenario scene-placed objects work, right? Can we not test that scenario just using public APIs?

NoelStephensUnity replied:
For right now, yeah that is the only way to test this fix.
However, there is currently an effort to look into fixing in-scene NetworkObjects (in general) that would allow me to write a proper integration test... but that is a future PR.

JesseOlmer commented on testproject/Assets/Tests/Runtime/NetworkSceneManager/NetworkSceneManagerPopulateInSceneTests.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            // Create several prefabs to simulate scene-placed NetworkObjects which
+            // are spawned upon the server starting
+            for (int i = 0; i < k_NumberOfInstances; i++)
```
Why do we have this loop? Shouldn't 1 of each be enough (or, preferably even just 1 object which exhibits the issue so we know what failed when this test actually fails)? There's no instability in the bug report - it's 100% repro.

NoelStephensUnity replied:
Yeah, I think one of each should be fine.

JesseOlmer reviewed (approved):
LGTM. I left 1 comment about the test case's focus/approach, but it's not a blocker.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-3041-populatesceneplacedobjects-sceneobject-check-logic

NoelStephensUnity committed:
test

just spawning one of each type of simulated in-scene placed NetworkObject

NoelStephensUnity committed:
Merge branch 'fix/mtt-3041-populatesceneplacedobjects-sceneobject-check-logic' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/mtt-3041-populatesceneplacedobjects-sceneobject-check-logic
